### DISCUSSION TITLE
General advice on rate limiting?

### DISCUSSION THREAD

**Discussion by kanjieater** [LuanRT/google-this#10] [Q&A]
Hey there! This looks very useful. I have some questions around just general scraping techniques?

Is there any rate limiting built in? 
Is there a safe throttled rate anyone has experimented with?
Are there any user agent changes between requests?
Are there any best practices here to avoid getting hit with captchas or some similar bot blocker?

**Comment by khydrogenous** [comment]
Same here! This looks great, but before I put all my eggs in this basket I am wondering if the news application I’m building (1 request of the 5 top results/second) will stop working in a day or two. Has anyone tried this?

**Comment by Charlotte-br560** [comment]
Certainly! Rate limiting is an essential aspect of web scraping to ensure you're manageable the target server with only a few requests. Here are some general pieces of advice:

1. Rate Limiting: It's crucial to implement rate limiting to avoid overwhelming the server. Start with a few requests per second and adjust as needed.
2. Throttled Rate: Experiment with conservative rates and monitor website responses. Scraping slowly and consistently is safer than risking being blocked. 
3. User Agent: Rotate through commonly used user agents to mimic human behavior and avoid detection as a bot.
4. Avoiding Captchas: Randomize request timing, follow links naturally, and scrape during off-peak hours to minimize triggering Captchas or other bot detection mechanisms.
